### PR TITLE
Refactor for Nova resource commonality

### PR DESCRIPTION
**Pull request by pdbreen**
Not yet ready for merge

### CODE REVIEW

**Review by drewroberts — COMMENTED** [review]
I like where you are heading with this...

File: src/Nova/Resource.php [hunk 1]
```diff
@@ -4,11 +4,94 @@
...
 
 abstract class Resource extends NovaResource
 {
+    protected array $filterClassList = [];
+
+    protected array $cardClassList = [];
+
+    protected array $lensClassList = [];
+
+    protected array $actionClassList = [];
```
Comment (pdbreen) [inline src/Nova/Resource.php:23]: Allow resources to declare filters, cards, lenses and actions that might be loaded in other classes by class name.

File: src/Nova/Resource.php [hunk 2]
```diff
@@ -4,11 +4,94 @@
...
+        // and return map class to instantiated instance
+        return array_map(
+            function (string $class) {
+            return new $class;
+        },
+            array_filter($classes, function (string $class) {
+                return class_exists($class);
+            })
+        );
+    }
```
Comment (pdbreen) [inline src/Nova/Resource.php]: For cards, actions, filters and lenses - run through the list of class names that this resource could use -- if they are defined.  Those that aren't defined, are filtered before instantiating and returning an array of those that do exist.

Comment (drewroberts) [reply src/Nova/Resource.php]: I like that!

File: src/Nova/Resource.php [hunk 3]
```diff
@@ -4,11 +4,94 @@
...
 
+use Illuminate\Http\Request;
+use Laravel\Nova\Fields\BelongsTo;
+use Laravel\Nova\Fields\DateTime;
+use Laravel\Nova\Fields\Field;
+use Laravel\Nova\Fields\ID;
 use Laravel\Nova\Http\Requests\NovaRequest;
 use Laravel\Nova\Resource as NovaResource;
 
 abstract class Resource extends NovaResource
```
Comment (drewroberts) [inline src/Nova/Resource.php]: Should we call this BaseResource instead of Resource? That would allow a similar naming convention with our BaseModel.

Comment (pdbreen) [reply src/Nova/Resource.php]: Your call.

File: src/Nova/Resource.php [hunk 4]
```diff
@@ -4,11 +4,94 @@
...
+     */
+    protected function updaterDataFields(): array
+    {
+        $userResource = nova('user');
+
+        return $userResource ? [
+            BelongsTo::make('Updated By', 'updater', $userResource)->exceptOnForms(),
+            DateTime::make('Updated At')->exceptOnForms(),
+        ] : [];
+    }
```
Comment (drewroberts) [inline src/Nova/Resource.php:59]: In every Nova resource, I like to separate out the following into a panel at the bottom:

- ID field
- Timestamps
- Creator (if applicable)
- Updater (if applicable)

Example:

```
protected function dataFields()
    {
        return [
            ID::make(),
            BelongsTo::make('Created By', 'creator', \App\Nova\User::class)->exceptOnForms(),
            DateTime::make('Created At')->exceptOnForms(),
            BelongsTo::make('Updated By', 'updater', \App\Nova\User::class)->exceptOnForms(),
            DateTime::make('Updated At')->exceptOnForms(),
        ];
    }
```

Could we find a way to add that to this resource stub?

Comment (pdbreen) [reply src/Nova/Resource.php:59]: You'll see the base class helpers are used in the derived classes to merge in the creator and updater fields when appropriate.

```
    protected function dataFields(): array
    {
        return array_merge(
            parent::dataFields(),
            $this->creatorDataFields(),
            $this->updaterDataFields(),
        );
    }
```

File: src/Nova/Resource.php (hunk 3 above)
Comment (drewroberts) [reply src/Nova/Resource.php]: OK, I prefer doing that then. I'll make the commit.

File: src/Nova/Resource.php (hunk 4 above)
Comment (drewroberts) [reply src/Nova/Resource.php:59]: Ok, thank you. Does this preserve the order with the timestamps?

Comment (pdbreen) [reply src/Nova/Resource.php:59]: Yes -- in this example, the end result is identical.

Comment (drewroberts) [reply src/Nova/Resource.php:59]: Thank you!

**Review by drewroberts — COMMENTED** [review]
Can we also pull this into the BaseResource?

`public static $model = \Tipoff\Checkout\Models\Order::class;`

**Review by drewroberts — APPROVED** [review]
Thank you! :shipit:

**Pull request merged**